README.txt
OVERVIEW
----------------------

Informa is an open source Java API for building applications which want to
access, modify and persist news channels information to allow content
syndication.

PLEASE NOTE: That although the Informa API is in a pretty stable state the
whole distribution is not complete and still considered in beta stadium.
Please report any problems to the informa-developers mailing list. Thanks
for your patience.

Additional information and documentation can be obtained from the Informa
web site: http://informa.sourceforge.net/


GETTING STARTED
----------------------

Requirements: it is assumed that you have a running Java 2 SDK (v. >= 1.3) 
environment and a SAX Parser (like Xerces-J) available (not needed if you
already use J2SDK 1.4 and higher).

Append the necessary JARs in the lib directory (especially informa.jar)
to your CLASSPATH. Together with the JavaDoc you should be able to get up to
speed in relative short time, one main goal in the design of the Informa
API was to be as much intuitive as possible.

This distribution will also includes an example application at some point in
the future. In the meantime please see the test directory in the src
package or in CVS to obtain samples and a minigui test application.


To verify that parsing is working properly you might want to execute for a
local feed:

  ant parse -Dargs='-f /path/to/my-weblog.xml'

If the feed is available from an URI please specify:

  ant parse -Dargs='-u http://url.to/my-weblog.xml'

Happy news channels parsing.


NOTES
----------------------

Please note that this release is a beta release.  Although Informa should be
fairly stable at this point, there is the possibility for changes before the
final release.

If you like you could have a look into the FAQ which is at:
http://informa.sourceforge.net/faq.html


CONTACT
----------------------

Please send all questions to niko_schmuck@users.sourceforge.net or join the
Informa developers list by visiting the web site
http://lists.sourceforge.net/mailman/listinfo/informa-developer .


Have fun,
Niko Schmuck


CHANGELOG.txt
Informa change log:

------------------------------------------------------------------------------
Version 0.7.0-optivo

 o Removed all System.out.println(...) and Exception.printStacktrace() calls
 
 o Changed all logger.info(...) calls to LOG.debug(...) calls

------------------------------------------------------------------------------
Version 0.7.0-alpha1

 o Minor code beautifications (JavaDoc)
 
 o Improved build system: 
   - two distrubtions á la Spring (with and without dependencies, sources
     and tests are included anyway)
   - switched from tar.gz to zip compression
   
 o Bringing Maven (1) project build back again for site generation

------------------------------------------------------------------------------
Version 0.7.0-alpha1

 o Improved parsing of Atom (0.3 and 1.0) feeds (thanks to Nilesh Bansal)
 
 o Migrated to JDK 5 (thanks to Ítalo Borssatto)

 o New PersistenceManager for better integration into the Spring Framework
   see de.nava.informa.utils.manager.hibernate.SpringPersistenceManager

 o Updated 3rd party libraries:
   - JDOM library from b10 to 1.0
   - Lucene from 1.2 to 1.4.3
   - Hibernate from 2.1 to 3.0.5 (and dependent libraries like Jakarta commons)
   - MySQL driver from 3.0.9 to 3.0.17
   - Log4J from 1.2.8 to 1.2.9

 o Removed unecessary 3rd party libraries:
   - XML parsing libraries, since we now assume at least JDK 1.4
   - PostgreSQL JDBC driver (you have to bring it with you on your own)
   - Castor JDO (they were outdated anyway, and the informa Castor JDO 
     binding experimental, the sources are still shipped anyway)

 o Removed xdoclet dependency for generating Hibernate mapping files (hbm)
   from now on we will maintain them manually

 o Removed implementation for the JDO backend, it was not maintained for a
   long time anyway
   
 o Fixed bug 1560571 (Exporting image element for the channel)
 
------------------------------------------------------------------------------
Version 0.6.5
 
 o Fixed bug 1145998 (Encoding not correctly set on writing feed)
   Thanks for reporting on this to Felipe Vieira Silva and Filip van Laenen.

 o [adewale@gmail.com, 01-May-2005] Fix for blogs (msdn) which have
   'created' dates, but not 'issued' dates in clear contravention of the Atom
   0.3 spec (Atom 0.3 Parser).

 o Changes by Aleksey Gureev (BlogBridge Team):
   - Added recognition of content type for Atom tags 'title', 'copyright',
     'content' and 'summary'. Title and copyright are returned in plain text.
     Content / summary are preferred in HTML format. Content tag is preferred
     source of entry (item) description.
   - Added support of <content:encoded> tag for RSS1.0 and RSS2.0.
   - Added support of muliple links in entries of Atom feed with different
     types. The one to be put in item object is selected using the order of
     type preference (currently it's set to 'text/html', then 'text/plain',
     then others.
   - Added support for another date format met at http://www.jnsk.se/weblog:
     "Tue, 21 Dec 2004 23:02 +0100".
   - Made poller and cleaner worker threads be daemons.
   - Reverted equals() matching methods to perform minimum comparisons for
     all implementations of ChannelIF and ItemIF (others don't have equals()).
   - Reworked Poller and Cleaner to process tasks basing on simple blocking
     queue, which is significantly more simple and reliable.
   - Added items scanning policies to Poller to control how it looks for the
     new items:
      POLICY_SCAN_ALL            - Scan all items (default)
      POLICY_SKIP_AFTER_EXISTING - Stop scanning once existing item detected.
   - Fixed problems with orphan guide available after successful merging.
   - Added Poller construction method with custom HTTP user-agent.
   - Added support for custom SAX2 Drivers in FeedParser.
   - Reworked session locking in HibernateUtil to improve thread-safety.
   - Added callback interface InputSourceProviderIF to convert feed's
     InputStream before parsing, which is useful for pre-processing the data
     before it gets into parser.
   - Converting case of tags' names to expected before feeds parsing.
   - Poller Module: Channels can be scheduled for polling with initial delay
     different from zero.
   - Poller Module: Added stream providing hook to create non-default
     InputStream for feed fetching.
   - Added detection of Atom 0.1-0.3 to FormaDetector.
   - Fix for invalid TTL format parsing in RSS 2.0.
   - Fix for parsing of Atom feeds without version attribute (v0.3 is assumed).
   - Added support for correct CDATA handling in the fields of Atom feeds
     (title, description, copyright, or whatever).
   - Integrated minor Hibernate + Channel/Item URL's fix by Scott Douglass

------------------------------------------------------------------------------
Version 0.6.0

 o upgraded CGLib to 2.0.2 (full)

 o Changes by Aleksey Gureev (BlogBridge Team):
   - Added universal Persistence Management interface. At the present moment
     interface supports operations over ChannelGroup's, Channel's and Item's.
   - Added In-Memory and Hibernate persistence manager implementations.
   - Added Poller module for convenient multi-threaded background polling of
     changes in channels.
   - Added Cleaner module for convenient multi-threaded background cleaning
     of channels using the set of rules.
   - Changed the way two URL's compared in implementations of ChannelIF and
     ItemIF to perform better and unleash the power of multi-threaded
     environment.

 o Fixed feature request 1015930 (long channelId vs. int channelId:
   ChannelRegistry)
 
 o Fixed feature request 1015926 (Issue Tracking link is bad)


------------------------------------------------------------------------------
Version 0.5.5

 o first experimental support for reading in Atom syntax (0.3 spec)
 
 o renamed de.nava.informa.parsers.RSSParser 
   to de.nava.informa.parsers.FeedParser to express that we

src/de/nava/informa/parsers/FeedParser.java
//
// Informa -- RSS Library for Java
// Copyright (c) 2002 by Niko Schmuck
//
// Niko Schmuck
// http://sourceforge.net/projects/informa
// mailto:niko_schmuck@users.sourceforge.net
//
// This library is free software.
//
// You may redistribute it and/or modify it under the terms of the GNU
// Lesser General Public License as published by the Free Software Foundation.
//
// Version 2.1 of the license should be included with this distribution in
// the file LICENSE. If the license is not included with this distribution,
// you may find a copy at the FSF web site at 'www.gnu.org' or 'www.fsf.org',
// or you may write to the Free Software Foundation, 675 Mass Ave, Cambridge,
// MA 02139 USA.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied waranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//


// $Id: FeedParser.java,v 1.13 2006/12/04 23:40:49 italobb Exp $

package de.nava.informa.parsers;

import de.nava.informa.core.ChannelBuilderIF;
import de.nava.informa.core.ChannelIF;
import de.nava.informa.core.ParseException;
import de.nava.informa.core.UnsupportedFormatException;
import de.nava.informa.utils.NoOpEntityResolver;
import de.nava.informa.utils.ParserUtils;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.jdom.Document;
import org.jdom.Element;
import org.jdom.JDOMException;
import org.jdom.Attribute;
import org.jdom.Namespace;
import org.jdom.input.SAXBuilder;
import org.xml.sax.InputSource;
import org.xml.sax.XMLReader;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.Reader;
import java.net.URL;

/**
 * Parser class which allows reading in of RSS news channels.
 * The concrete rules how the XML elements map to our channel object model
 * are delegated to version specific private classes.</p>
 * <p/>
 * Currently the FeedParser support RSS formats 0.9x, 1.0 (RDF), 2.0 and Atom 0.3.
 * <p/>
 * It's possible to specify class of custom XML parser to use instead of
 * standard Crimson parser (JDK). Use <code>setSaxDriverClassName()</code> method
 * for that.
 *
 * @author Niko Schmuck
 */
public class FeedParser {

    private static final Log LOG = LogFactory.getLog(FeedParser.class);

    private static String saxDriverClassName = null;

    /**
     * Sets the name of SAX2 Driver class to use for parsing. The class should
     * implement XMLReader interface and should exist in current class-loading
     * paths. The method will check these conditions before accepting specified
     * class name.
     *
     * @param className name of SAX2 Driver class.
     * @throws ClassNotFoundException if class isn't found in the class path.
     * @throws ClassCastException     if class isn't implementing XMLReader interface.
     */
    public static synchronized void setSaxDriverClassName(String className) throws ClassNotFoundException {

        // Check if class is available
        Class saxDriverClass = Class.forName(className);

        // Check if class implements necessary XMLReader interface
        Class[] extendsImplements = saxDriverClass.getInterfaces();
        boolean found = false;
        for (int i = 0; !found && i < extendsImplements.length; i++) {
            Class parent = extendsImplements[i];
            found = (parent == XMLReader.class);
        }
        if (!found) throw new ClassCastException("Specified class " + className +
                " does not implement XMLReader.");

        saxDriverClassName = className;
    }

    /**
     * Parser feed behind the given URL and build channel.
     *
     * @param cBuilder specific channel builder to use.
     * @param url      URL to use as data source.
     * @return parsed channel.
     * @throws IOException    if IO errors occur.
     * @throws ParseException if parsing is not possible.
     */
    public static ChannelIF parse(ChannelBuilderIF cBuilder, String url) throws IOException, ParseException {
        final URL aURL;
        try {
            aURL = new URL(url);
        } catch (java.net.MalformedURLException e) {
            throw new IOException("Invalid url '" + url + "' had invalid URL representation.");
        }
        return parse(cBuilder, new InputSource(url), aURL);
    }

    /**
     * Parser feed behind the given URL and build channel.
     *
     * @param cBuilder specific channel builder to use.
     * @param aURL     URL to use as data source.
     * @return parsed channel.
     * @throws IOException    if IO errors occur.
     * @throws ParseException if parsing is not possible.
     */
    public static ChannelIF parse(ChannelBuilderIF cBuilder, URL aURL) throws IOException, ParseException {
        return parse(cBuilder, new InputSource(aURL.toExternalForm()), aURL);
    }

    /**
     * Parse feed presented by Reader and build channel.
     *
     * @param cBuilder specific channel builder to use.
     * @param reader   reader setup to read feed data.
     * @return parsed channel.
     * @throws IOException    if IO errors occur.
     * @throws ParseException if parsing is not possible.
     */
    public static ChannelIF parse(ChannelBuilderIF cBuilder, Reader reader) throws IOException, ParseException {
        return parse(cBuilder, new InputSource(reader), null);
    }

    /**
     * Parse feed presented by InputStream and build channel.
     *
     * @param cBuilder specific channel builder to use.
     * @param stream   stream setup to read feed data.
     * @return parsed channel.
     * @throws IOException    if IO errors occur.
     * @throws ParseException if parsing is not possible.
     */
    public static ChannelIF parse(ChannelBuilderIF cBuilder, InputStream stream) throws IOException, ParseException {
        return parse(cBuilder, new InputSource(stream), null);
    }

    /**
     * Parse feed presented by file and build channel.
     *
     * @param cBuilder specific channel builder to use.
     * @param aFile    file to read data from.
     * @return parsed channel.
     * @throws IOException    if IO errors occur.
     * @throws ParseException if parsing is not possible.
     */
    public static ChannelIF parse(ChannelBuilderIF cBuilder, File aFile) throws IOException, ParseException {
        final URL aURL;
        try {
            aURL = aFile.toURI().toURL();
        } catch (java.net.MalformedURLException e) {
            throw new IOException("File " + aFile + " had invalid URL representation.");
        }
        return parse(cBuilder, new InputSource(aURL.toExternalForm()), aURL);
    }

    /**
     * Parse feed from input source with base location set and create channel.
     *
     * @param cBuilder     specific channel builder to use.
     * @param inpSource    input source of data.
     * @param baseLocation base location of feed.
     * @return parsed channel.
     * @throws IOException    if IO errors occur.
     * @throws ParseException if parsing is not possible.
     */
    public static ChannelIF parse(ChannelBuilderIF cBuilder, InputSource inpSource,
                                  URL baseLocation)
            throws IOException, ParseException {
        // document reading without validation
        SAXBuilder saxBuilder = new SAXBuilder(saxDriverClassName);

        // turn off DTD loading
        saxBuilder.setEntityResolver(new NoOpEntityResolver());

        try {
            Document doc = saxBuilder.build(inpSource);
            ChannelIF channel = parse(cBuilder, doc);
            channel.setLocation(baseLocation);
            return channel;
        } catch (JDOMException e) {
            throw new ParseException("Problem parsing " + inpSource.getSystemId() + ": " + e);
        }
    }

    // ------------------------------------------------------------
    // internal helper methods
    // ------------------------------------------------------------

    private static synchronized ChannelIF parse(ChannelBuilderIF cBuilder, Document doc) throws ParseException {
        if (cBuilder == null) {
            throw new RuntimeException("Without builder no channel can be created.");
        }
        if (LOG.isDebugEnabled()) {
            LOG.debug("start parsing.");
        }
        // Get the root element (must be rss)
        Element root = doc.getRootElement();
        String rootElement = root.getName().toLowerCase();
        // Decide which parser to use
        if (rootElement.startsWith("rss")) {
            String rssVersion = root.getAttribute("version").getValue();
            if (rssVersion.indexOf("0.91") >= 0) {
                if (LOG.isDebugEnabled()) {
                    LOG.debug("Channel uses RSS root element (Version 0.91).");
                }
                return RSS_0_91_Parser.getInstance().parse(cBuilder, root);
            } else if (rssVersion.indexOf("0.92") >= 0) {
                if (LOG.isDebugEnabled()) {
                    LOG.debug("Channel uses RSS root element (Version 0.92).");
                }
                // logger.warn("RSS 0.92 not fully supported yet, fall back to 0.91.");
                // TODO: support RSS 0.92 when aware of all subtle differences.
                return RSS_0_91_Parser.getInstance().parse(cBuilder, root);
            } else if (rootElement.indexOf("0.93") >= 0) {
                if (LOG.isDebugEnabled()) {
                    LOG.debug("Channel uses RSS root element (Version 0.93).");
                }
                LOG.info("RSS 0.93 not fully supported yet, fall back to 0.91.");
                // TODO: support RSS 0.93 when aware of all subtle differences.
            } else if (rootElement.indexOf("0.94") >= 0) {
                if (LOG.isDebugEnabled()) {
                    LOG.debug("Channel uses RSS root element (Version 0.94).");
                }
                LOG.info("RSS 0.94 not fully supported yet, will use RSS 2.0");
                // TODO: support RSS 0.94 when aware of all subtle differences.
                return RSS_2_0_Parser.getInstance().parse(cBuilder, root);
            } else if (rssVersion.indexOf("2.0") >= 0 || rssVersion.equals("2")) {
                if (LOG.isDebugEnabled()) {
                    LOG.debug("Channel uses RSS root element (Version 2.0).");
                }
                return RSS_2_0_Parser.getInstance().parse(cBuilder, root);
            } else {
                throw new UnsupportedFormatException("Unsupported RSS version [" + rssVersion + "].");
            }
        } else if (rootElement.indexOf("rdf") >= 0) {
            return RSS_1_0_Parser.getInstance().parse(cBuilder, root);
        } else if (rootElement.indexOf("feed") >= 0) {
            Attribute versionAttr = root.getAttribute("version");
            Namespace namespace = ParserUtils.getDefaultNS(root);
            if (versionAttr != null) {
                String feedVersion = versionAttr.getValue();
                if (feedVersion.indexOf("0.1") >= 0 || feedVersion.indexOf("0.2") >= 0) {
                    if (LOG.isDebugEnabled()) {
                        LOG.debug("Channel uses feed root element (Version " + feedVersion + ").");
                    }
                    LOG.info("This atom version is not really supported yet, assume Atom 0.3 format");
                    return Atom_0_3_Parser.getInstance().parse(cBuilder, root);
                } else if (feedVersion.indexOf("0.3") >= 0) {
                    if (LOG.isDebugEnabled()) {
                        LOG.debug("Channel uses feed root element (Version 0.3).");
                    }
                    return Atom_0_3_Parser.getInstance().parse(cBuilder, root);
                }
            } else if (namespace != null && namespace.getURI() != null) {
                if (!"http://www.w3.org/2005/Atom".equals(namespace.getURI())) {
                    LOG.info("Channel uses unknown namespace in feed root element, assume Atom 1.0 format.");
                } else if (LOG.isDebugEnabled()) {
                    LOG.debug("Channel uses feed root element (Atom 1.0 format).");
                }
                return Atom_1_0_Parser.getInstance().parse(cBuilder, root);
            }
        }
        // did not match anything
        throw new UnsupportedFormatException("Unsupported root element [" + rootElement + "].");
    }
}